### ISSUE TITLE
Command '['gsnap', '...']' returned non-zero exit status 255.

### ISSUE THREAD

**Issue by sarahbonnin** [nedialkova-lab/mim-tRNAseq#4]
Hello,

I am trying to use mimseq and getting an error I don't manage to understand... It seems to fail at the "gsnap" step.

mimseq is run as follows as a job in an HTC cluster (usearch is added to the PATH at the start of the script, and rights were changed so it is executable by any user):

mimseq --species Hsap38 \
        --trnas GtRNAdb/hg38-tRNAs.fa \
        --trnaout GtRNAdb/hg38-tRNAs-detailed.out \
        --cluster \
        --threads 8 \
        --control-condition WT \
        --name mimiseq_test \
        --out-dir test \
        sample_data_mimseq.txt

I got the tRNA files from GtRNAdb. 

The sample_data_mimseq.txt and fastq.gz files can be found by the program (I first had that error and then changed the path and it was solved).

Here follows the error log (I have hidden the full paths for privacy reasons):

2021-02-25 10:51:09,723 [INFO ] mim-tRNAseq v0.3.3 run with command:
2021-02-25 10:51:09,724 [INFO ] /...path.../miniconda3/envs/mimseq/bin/mimseq 
--species Hsap38 --trnas GtRNAdb/hg38-tRNAs.fa --out hg38-tRNAs-detailed.out --clust
er --threads 8 --control-condition WT --name mimiseq_test--out-dir test sample_
data_mimseq.txt
2021-02-25 10:51:09,836 [INFO ] 
+----------------------------------------------+                
| Starting analysis for hg38-tRNAs-filtered.fa |                
+----------------------------------------------+
2021-02-25 10:51:09,836 [INFO ] Processing tRNA sequences...
2021-02-25 10:51:10,207 [INFO ] 34 introns registered...
2021-02-25 10:51:10,367 [INFO ] 600 cytosolic and 22 mitochondrial tRNA sequences im
ported
2021-02-25 10:51:10,367 [INFO ] Processing modomics database...
2021-02-25 10:51:13,522 [INFO ] Modomics retrieved...
2021-02-25 10:51:13,523 [INFO ] Parsing Modomics JSON data...
2021-02-25 10:51:13,531 [INFO ] Number of Modomics entries for Escherichia coli: 43
2021-02-25 10:51:13,531 [INFO ] Number of Modomics entries for Homo sapiens: 45
2021-02-25 10:51:33,038 [INFO ] 
+------------------------+              
| Beginning SNP indexing |              
+------------------------+
2021-02-25 11:00:01,011 [INFO ] 622 total tRNA gene sequences (undetermined and nmt 
sequences excluded)
2021-02-25 11:00:01,012 [INFO ] 65 sequences with a match to Modomics dataset
2021-02-25 11:00:01,022 [INFO ] **** Clustering tRNA sequences ****
2021-02-25 11:00:01,022 [INFO ] Clustering tRNA sequences by 97% similarity...
2021-02-25 11:00:16,441 [INFO ] 325 clusters created from 622 tRNA sequences
2021-02-25 11:00:16,448 [INFO ] 217 modifications written to SNP index
2021-02-25 11:00:16,448 [INFO ] 49 A to G replacements in reference sequences for inosine modifications
2021-02-25 11:00:18,005 [INFO ] 
+--------------------------+             
| Generating GSNAP indices |             
+--------------------------+
2021-02-25 11:00:49,671 [INFO ] Genome indices done...
2021-02-25 11:00:49,672 [WARNI] SNP-tolerant alignment turned off or no modifications found for input tRNAs: SNP indices not built...
2021-02-25 11:00:49,672 [INFO ] 
+-----------+                   
| Alignment |                   
+-----------+
2021-02-25 11:00:49,675 [INFO ] **** WT1_trimmed.fastq.gz ****
2021-02-25 11:00:49,675 [INFO ] Aligning reads to Hsap38_tRNAgenome...
Traceback (most recent call last):
  File "/...path.../miniconda3/envs/mimseq/bin/mimseq", line 10, in <module>
    sys.exit(main())
  File "/...path.../miniconda3/envs/mimseq/lib/python3.7/site-packages/mimseq/mimseq.py", line 361, in main
    args.misinc_thresh, args.mito, args.pretrnas, args.local_mod, args.sampledata)
  File "/...path.../miniconda3/envs/mimseq/lib/python3.7/site-packages/mimseq/mimseq.py", line 96, in mimseq
    snp_index_path, snp_index_name, out, threads, snp_tolerance, keep_temp, mismatches, map_round)
  File "/...path.../miniconda3/envs/mimseq/lib/python3.7/site-packages/mimseq/tRNAmap.py", line 48, in mainAlign
    unique_bam, librarySize, alignstats = mapReads(fq, genome_index_path, genome_index_name, snp_index_path, snp_index_name, threads, out_dir, snp_tolerance, keep_temp, mismatches, remap)
  File "/...path.../miniconda3/envs/mimseq/lib/python3.7/site-packages/mimseq/tRNAmap.py", line 217, in mapReads
    subprocess.check_call(map_cmd, stderr = open(out_dir + "align.log", "a"))
  File "/...path.../miniconda3/envs/mimseq/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['gsnap', '--gunzip', '-D', 'test/Hsap38_tRNAgenome', '-d', 'Hsap38_tRNAgenome', '-t', '8', '--split-output', 'test/WT1_trimmed', '--format', 'sam', '--genome-unk-mismatch', '0', '--md-lowercase-snp', '--ignore-trim-in-filtering', '1', '/...path.../fastq_trimmed/WT1_trimmed.fastq.gz']' returned non-zero exit status 255.

I'm not sure what I am doing wrong! 

Any tip is welcome: thank you so much!

Best regards,
Sarah

**Comment by drewjbeh** [comment]
Hi Sarah,

Thanks fo using mimseq and reporting the error. Sorry you are having issues getting it to work!
Firstly I would like to say that if you specify `--species Hsap38` you don't need to supply a fasta file and "out" file. These are included automatically when you specify `--species`. These options can be specified if you have a custom reference set you would like to align to, but if you would just like to align to hg38, I recommend just using `--species Hsap38` and nothing else.

For the actual error, you should see an `align.log` file in the output directory. Could you have a look at that and share with me, there will probably be a hint in there as to what's going wrong!

Thanks,
Drew

**Comment by sarahbonnin** [comment]
Hi Drew,

Thank you for the very prompt reply!

Well noted, for the `--species Hsap38` option!
I was doubting whether to get the files or not because, in mimseq's data folder, I don't see the hg38-tRNA folder, only the hg38-eColitK:

`ls ~/miniconda3/envs/mimseq/pkgs/mimseq-0.3.3-py_0/site-packages/mimseq/data/`

additionalMods_all.txt       eschColi-tRNA_Lys            modomics_orig
additionalMods_original.txt  hg19-eColitK                 sacCer3-eColitK
additionalMods.txt           hg19-eColitK-highConf-tRNAs  sacCer3_modOligos
ce11-tRNAs                   hg19-tRNAs                   sacCer3-Phe_SynVsPur
danRer11-eColitK             hg38-eColitK                 sacCer3-tRNAs
dm6-eColitK                  mm10-eColitK                 schiPomb-972H
dm6-tRNAs                    mm10-tRNAs                   schiPomb-eColitK
eschColi-K_12_MG1655-tRNAs   modomics                     tRNAmatureseq.cm

Is it normal?

Going back to the first issue: the `align.log` file exists, but it is actually empty...

Cheers

**Comment by drewjbeh** [comment]
Hi Sarah,

Yes, you are correct that there is only a hg38-eColitK folder. This reference just contains the additional E. coli Lys-TTT reference we used in the paper as a spike-in. It should be completely ok to use this reference if you did not generate your libraries with this spike-in, as no reads will align here and it will essentially be excluded from the results. 

Hmm, interesting that it is empty. Could you try run `gsnap --version` just to check that gsnap is indeed installed and available. Also the version should be "2019-02-26". I'm thinking that if mimseq tries to call gsnap and it can't find that command in the PATH, then it would throw the error and not even generate anything for the log file.

**Comment by sarahbonnin** [comment]
Hi,

gsnap is found in the conda environment, version 2019-02-26 indeed:

`gsnap --version`

GSNAP version 2019-02-26 called with args: gsnap.sse42 --version

GSNAP: Genomic Short Nucleotide Alignment Program
Part of GMAP package, version 2019-02-26
Build target: x86_64-unknown-linux-gnu
Features: pthreads enabled, no alloca, zlib available, mmap available, littleendian, sigaction available, 64 bits available
Popcnt: mm_popcnt builtin_popcount
Builtin functions: builtin_clz builtin_ctz builtin_popcount
SIMD functions compiled: SSE2 SSSE3 SSE4.1 SSE4.2
Sizes: off_t (8), size_t (8), unsigned int (4), long int (8), long long int (8)
Default gmap directory (compiled): /...path.../miniconda3/envs/mimseq/share
Default gmap directory (environment): /...path.../miniconda3/envs/mimseq/share
Maximum stack read length: 300
Thomas D. Wu, Genentech, Inc.
Contact: twu@gene.com

(mimseq) [jue feb 25 12:30:56, 0 jobs, ---@---: /...path.../test ]$

`which gsnap`
~/miniconda3/envs/mimseq/bin/gsnap

**Comment by drewjbeh** [comment]
Could you try run the gsnap step manually. So assemble a command from `['gsnap', '--gunzip', '-D', 'test/Hsap38_tRNAgenome', '-d', 'Hsap38_tRNAgenome', '-t', '8', '--split-output', 'test/WT1_trimmed', '--format', 'sam', '--genome-unk-mismatch', '0', '--md-lowercase-snp', '--ignore-trim-in-filtering', '1', '/...path.../fastq_trimmed/WT1_trimmed.fastq.gz']` by removing [ ] and the quotes and pointing gsnap to correct paths of the indices built by mimseq already and the fastq file

**Comment by sarahbonnin** [comment]
Hi,

The command line doesn't work when I send it as a job to the HPC cluster, but it **does** work when I run it locally on the server: exact same PATH and conda environment etc., exact same working directory...
I don't know what is happening, but it's obviously an issue that I have here, and it doesn't seem related to the tool at the end.

I will try and figure it out: thank you so much for your help, and sorry for wasting your time!

Best,
Sarah

**Comment by drewjbeh** [comment]
Hi Sarah,

I see! It might have something to do with submitting to the cluster...I will need to investigate this and see what gsnap is able to do in this case (along with all other tools and code in the package). Unfortunately, I didn't write mimseq with clusters in mind :(. I'll probably add this specifically to the README or something. Plus, its good to know for future functionality upgrades, so definitely not a waste of my time at all!

Is it possible for you to run the whole mimseq pipeline on your local server? It's really not too computationally intensive, especially if you reduce `--threads` and `--max-multi`, and shouldn't take too long either. For example with 40 threads (I know, quite a lot), and `--max-multi 6` on 10 samples of between 3 and 7 million reads in human took only 30 minutes yesterday afternoon.

**Comment by sarahbonnin** [comment]
Hi Drew,

Thanks for your advice!

I submitted the job to different queues in the computer cluster, and it seems that, in some queues, gsnap doesn't work properly. I'm thinking it might be a problem with the conda environment set up in some of the queues, that's the only thing I can think about... I will report it to our IT and see if they can make an upgrade.

In any case: I located a queue where the script runs well, and the job is now ongoing, and it didn't fail at the gsnap step!
So it looks like your tool can be run in the cluster, and that we have issues here with some of the queues.

Thanks!

**Comment by drewjbeh** [comment]
Hi Sarah,

Ah great! Good to know. Glad it's working in at least some queues. Let me know if you run into other issues.

**Comment by sherry-mm** [comment]
Hi sarahbonnin！
May I ask how you finally resolved this issue? It seems that we both encountered the same problem:
（1）I also encountered the following error message: 
subprocess.CalledProcessError: Command '['gsnap', '--gunzip', '-D', 'mimseq_HEKvsK562_results/Hsap_tRNAgenome', '-d', 'Hsap_tRNAgenome', '-V', 'mimseq_HEKvsK562_results/Hsapsnp_index', '-v', 'HEK_vs_K562_demo_modificationSNPs', '-t', '8', '--split-output', 'mimseq_HEKvsK562_results/mimseq_hek_1', '--format', 'sam', '--genome-unk-mismatch', '0', '--md-lowercase-snp', '--ignore-trim-in-filtering', '1', './mimseq_hek_1.fastq.gz']' returned non-zero exit status 255.
（2）Moreover, when I ran the "gsnap --version" command, no information appeared.
 However, when I ran "gsnap.sse42 --version", it showed: GSNAP:
Genomic Short Nucleotide Alignment Program
Part of GMAP package, version 2019-02-26
Build target: x86_64-unknown-linux-gnu
Features: pthreads enabled, no alloca, zlib available, mmap available, littleendian, sigaction available, 64 bits available
Popcnt: mm_popcnt builtin_popcount
Builtin functions: builtin_clz builtin_ctz builtin_popcount
Compiled SIMD functions: SSE2, SSSE3, SSE4.1, SSE4.2 Sizes: off_t (8), size_t (8), unsigned int (4), long int (8), long long int (8)
Default gmap directory (compiled):  /... path... /miniconda3/envs/mimseq/share
Default gmap directory (environment):  /... path... /miniconda3/envs/mimseq/share
Maximum stack read length: 300
Thomas D. Wu, Genentech, Inc.
Contact: twu@gene.com
I'm looking forward to your reply. Thank you!❤
